content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,198 @@
+{
+        "type": "c",
+        "path": "/dev/foo",
+        "major": 123,
+        "minor": 4,
+        "fileMode": 438,
+        "uid": <runAsUser>,
+        "gid": <runAsGroup>
+},
+```
+
+With `runc` OCI runtime (in non-[*rootless mode*](/docs/tasks/administer-cluster/kubelet-in-userns/)), the device is created (`mknod(2)`) in
+the container name space and the ownership is changed to `runAsUser`/`runAsGroup` using `chmod(2)` (NB: rootless mode and devices is currently not supported).
+Having the ownership updated in the container namespace is justified as the user process is the only one accessing the device. Only `runAsUser`/`runAsGroup`
+are taken into account, and, e.g., the `USER` setting in the container is currently ignored.
+
+While it is likely that the "faulty" deployments (i.e., non-root `securityContext` + devices) do not exist, to be absolutely sure no
+deployments break, an opt-in config entry in both containerd and CRI-O to enable the new behavior was added. The following:
+
+`device_ownership_from_security_context (bool)`
+
+defaults to `false` and must be enabled to use the feature.
+
+## See non-root containers using devices after the fix
+To demonstrate the new behavior, let's use a Data Plane Development Kit (DPDK) application using hardware accelerators, Kubernetes CPU manager, and HugePages as an example. The cluster runs containerd with:
+
+```toml
+[plugins]
+  [plugins."io.containerd.grpc.v1.cri"]
+    device_ownership_from_security_context = true
+```
+
+or CRI-O with:
+
+```toml
+[crio.runtime]
+device_ownership_from_security_context = true
+```
+
+and the `Guaranteed` QoS Class Pod that runs DPDK's crypto-perf test utility:
+
+```yaml
+...
+metadata:
+  name: qat-dpdk
+spec:
+  securityContext:
+    runAsUser: 1000
+    runAsGroup: 2000
+    fsGroup: 3000
+  containers:
+  - name: crypto-perf
+    image: intel/crypto-perf:devel
+    ...
+    resources:
+      requests:
+        cpu: "3"
+        memory: "128Mi"
+        qat.intel.com/generic: '4'
+        hugepages-2Mi: "128Mi"
+      limits:
+        cpu: "3"
+        memory: "128Mi"
+        qat.intel.com/generic: '4'
+        hugepages-2Mi: "128Mi"
+  ...
+```
+
+`fsGroup` gives a `runAsUser` writable HugePages emptyDir mountpoint. To verity the results, check the container's environment:
+
+```
+$ kubectl exec -it qat-dpdk -c crypto-perf -- id
+uid=1000 gid=2000 groups=2000,3000
+$ kubectl exec -it qat-dpdk -c crypto-perf -- ls -la /dev/vfio
+total 0
+drwxr-xr-x 2 root root      140 Sep  7 10:55 .
+drwxr-xr-x 7 root root      380 Sep  7 10:55 ..
+crw------- 1 1000 2000 241,   0 Sep  7 10:55 58
+crw------- 1 1000 2000 241,   2 Sep  7 10:55 60
+crw------- 1 1000 2000 241,  10 Sep  7 10:55 68
+crw------- 1 1000 2000 241,  11 Sep  7 10:55 69
+crw-rw-rw- 1 1000 2000  10, 196 Sep  7 10:55 vfio
+$ kubectl exec -it qat-dpdk -c crypto-perf -- ls -la /dev/hugepages/
+total 0
+drwxrwsr-x 2 root 3000   0 Sep  7 10:55 .
+drwxr-xr-x 7 root root 380 Sep  7 10:55 ..
+```
```

Comment:
Something like this? 

One-line description of what the command does:

```shell
kubectl exec -it qat-dpdk -c crypto-perf -- id
```

The output is similar to the following:
```
uid=1000 gid=2000 groups=2000,3000
```

One-line description of what the command does:
```shell
kubectl exec -it qat-dpdk -c crypto-perf -- ls -la /dev/vfio
```

The output is similar to the following:
```
total 0
drwxr-xr-x 2 root root      140 Sep  7 10:55 .
drwxr-xr-x 7 root root      380 Sep  7 10:55 ..
crw------- 1 1000 2000 241,   0 Sep  7 10:55 58
crw------- 1 1000 2000 241,   2 Sep  7 10:55 60
crw------- 1 1000 2000 241,  10 Sep  7 10:55 68
crw------- 1 1000 2000 241,  11 Sep  7 10:55 69
crw-rw-rw- 1 1000 2000  10, 196 Sep  7 10:55 vfio
```

One-line description of what the command does:
```shell
kubectl exec -it qat-dpdk -c crypto-perf -- ls -la /dev/hugepages/
```

The output is similar to the following:
```
total 0
drwxrwsr-x 2 root 3000   0 Sep  7 10:55 .
drwxr-xr-x 7 root root 380 Sep  7 10:55 ..
```

Reply:
@shannonxtreme thanks!

Reply:
fixed